src/pd/async/util.rs

```diff
@@ -0,0 +1,296 @@
+                    }
+                };
+                this
+            })
+            .boxed()
+    }
+
+    fn check(self) -> PdFuture<(GetClient<Req, Resp, F>, bool)> {
+        if self.retry_count == 0 {
+            return ok((self, true)).boxed();
+        }
+
+        if self.resp.as_ref().map_or(false, |ret| ret.is_ok()) {
+            debug!("GetClient get Some(Ok(_))");
+            return ok((self, true)).boxed();
+        }
+
+        // FIXME: should not block the core.
+        warn!("updating PD client, block the tokio core");
+
+        let start = Instant::now();
+        // Go to sync world.
+        let members = self.bundle.rl().members.clone();
+        match try_connect_leader(&members) {
+            Ok((client, members)) => {
+                let mut bundle = self.bundle.wl();
+                if members != bundle.members {
+                    bundle.client = Arc::new(client);
+                    bundle.members = members;
+                }
+                warn!("updating PD client done, spent {:?}", start.elapsed());
+            }
+
+            Err(err) => {
+                warn!("updating PD client spent {:?}, err {:?}",
+                      start.elapsed(),
+                      err);
+                // FIXME: use tokio-timer instead.
+                thread::sleep(Duration::from_secs(1));
+            }
+        }
+
+        ok((self, false)).boxed()
+    }
+
+    fn get_resp(self) -> Option<Result<Resp>> {
+        self.resp
+    }
+
+    /// Returns a Future, it is resolves once a future returned by the closure
+    /// is resolved successfully, otherwise it repeats `retry` times.
+    pub fn reconnect(self) -> PdFuture<Resp> {
+        let this = self;
+        loop_fn(this, |this| {
+                this.get()
+                    .and_then(|this| this.check())
+                    .and_then(|(this, done)| {
+                        if done {
+                            Ok(Loop::Break(this))
+                        } else {
+                            Ok(Loop::Continue(this))
+                        }
+                    })
+            })
+            .then(|req| {
+                match req.unwrap().get_resp() {
+                    Some(Ok(resp)) => future::ok(resp),
+                    Some(Err(err)) => future::err(err),
+                    None => future::err(box_err!("GetClient fail to request")),
+                }
+            })
+            .boxed()
+    }
+}
+
+struct GetClientRead<Req, Resp, F> {
+    inner: Option<GetClient<Req, Resp, F>>,
+}
+
+// TODO: impl Stream instead.
+impl<Req, Resp, F> Future for GetClientRead<Req, Resp, F> {
+    type Item = (GetClient<Req, Resp, F>, Arc<PDAsyncClient>);
+    type Error = Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        let inner = self.inner.take().expect("GetClientRead cannot poll twice");
+        let ret = match inner.bundle.try_read() {
```

Comment:
> Implementors of this function should ensure that a call to this never blocks as event loops may not work properly otherwise.

Updating client takes a write lock, if we don't use try read lock, it blocks.